ResidualAst mutates input *Ast

**Describe the bug**
`ResidualAst` mutates the input `*Ast` which causes subsequent evaluations to fail or return incorrect results.

**To Reproduce**
Check which components this affects:

- [ ]  parser
- [ ]  checker
- [x]  interpreter

Sample expression and input that reproduces the issue:

Test setup:
```go
func Test_ResidualAst_Modified(t *testing.T) {
	e, _ := cel.NewEnv(
		cel.Declarations(
			decls.NewVar("x", decls.NewMapType(decls.String, decls.Int)),
			decls.NewVar("y", decls.Int),
		),
	)
	ast, _ := e.Parse("x == y")
	prg, _ := e.Program(ast,
		cel.EvalOptions(cel.OptTrackState, cel.OptPartialEval),
	)
	for _, x := range []int{123, 456} {
		vars, _ := cel.PartialVars(map[string]interface{}{
			"x": x,
		}, cel.AttributePattern("y"))
		out, det, err := prg.Eval(vars)
		if !types.IsUnknown(out) {
			t.Fatalf("got %v, expected unknown", out)
		}
		if err != nil {
			t.Fatal(err)
		}
		residual, err := e.ResidualAst(ast, det)
		if err != nil {
			t.Fatal(err)
		}
		orig, err := cel.AstToString(ast)
		if err != nil {
			t.Fatal(err)
		}
		if orig != "x == y" {
			t.Errorf("parsed ast: got expr: %s, wanted x == y", orig)
		}
		expr, err := cel.AstToString(residual)
		if err != nil {
			t.Fatal(err)
		}
		want := fmt.Sprintf("%d == y", x)
		if expr != want {
			t.Errorf("residual ast: got expr: %s, wanted %s", expr, want)
		}
	}
}
```

```
=== RUN   Test_ResidualAst_Modified
    filter_test.go:83: parsed ast: got expr: 123 == y, wanted x == y
    filter_test.go:83: parsed ast: got expr: 123 == y, wanted x == y
    filter_test.go:91: residual ast: got expr: 123 == y, wanted 456 == y
--- FAIL: Test_ResidualAst_Modified (0.00s)
```

**Expected behavior**
`ResidualAst` should not be mutating it's input `*Ast` and should instead return a copy.

**Additional context**
~Either `ResidualAst` or `interpreter.PruneAst` should be making a copy of the expr that can then be safely mutated.~

--- a/cel/cel_test.go
+++ b/cel/cel_test.go
@@ -1068,3 +1068,47 @@ func Test_ResidualAst_AttributeQualifiers(t *testing.T) {
 		t.Errorf("got expr: %s, wanted %s", expr, want)
 	}
 }
+
+func Test_ResidualAst_Modified(t *testing.T) {
+	e, _ := NewEnv(
+		Declarations(
+			decls.NewVar("x", decls.NewMapType(decls.String, decls.Int)),
+			decls.NewVar("y", decls.Int),
+		),
+	)
+	ast, _ := e.Parse("x == y")
+	prg, _ := e.Program(ast,
+		EvalOptions(OptTrackState, OptPartialEval),
+	)
+	for _, x := range []int{123, 456} {
+		vars, _ := PartialVars(map[string]interface{}{
+			"x": x,
+		}, AttributePattern("y"))
+		out, det, err := prg.Eval(vars)
+		if !types.IsUnknown(out) {
+			t.Fatalf("got %v, expected unknown", out)
+		}
+		if err != nil {
+			t.Fatal(err)
+		}
+		residual, err := e.ResidualAst(ast, det)
+		if err != nil {
+			t.Fatal(err)
+		}
+		orig, err := AstToString(ast)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if orig != "x == y" {
+			t.Errorf("parsed ast: got expr: %s, wanted x == y", orig)
+		}
+		expr, err := AstToString(residual)
+		if err != nil {
+			t.Fatal(err)
+		}
+		want := fmt.Sprintf("%d == y", x)
+		if expr != want {
+			t.Errorf("residual ast: got expr: %s, wanted %s", expr, want)
+		}
+	}
+}

--- a/interpreter/planner.go
+++ b/interpreter/planner.go
@@ -232,15 +232,15 @@ func (p *planner) planSelect(expr *exprpb.Expr) (Interpretable, error) {
 		return attr, err
 	}
 
-	relAttr := p.attrFactory.RelativeAttribute(op.ID(), op)
+	relAttr, err := p.relativeAttr(op.ID(), op)
+	if err != nil {
+		return nil, err
+	}
 	_, err = relAttr.AddQualifier(qual)
 	if err != nil {
 		return nil, err
 	}
-	return &evalAttr{
-		adapter: p.adapter,
-		attr:    relAttr,
-	}, nil
+	return relAttr, nil
 }
 
 // planCall creates a callable Interpretable while specializing for common functions and invocation
@@ -477,20 +477,9 @@ func (p *planner) planCallIndex(expr *exprpb.Expr,
 	args []Interpretable) (Interpretable, error) {
 	op := args[0]
 	ind := args[1]
-	opAttr, isOpAttr := op.(InterpretableAttribute)
-	if !isOpAttr {
-		opAttr = &evalAttr{
-			adapter: p.adapter,
-			attr:    p.attrFactory.RelativeAttribute(op.ID(), op),
-		}
-	}
-	decAttr, err := p.decorate(opAttr, nil)
+	opAttr, err := p.relativeAttr(op.ID(), op)
 	if err != nil {
 		return nil, err
-	}
-	opAttr, isOpAttr = decAttr.(InterpretableAttribute)
-	if !isOpAttr {
-		return nil, fmt.Errorf("invalid attribute decoration: %v(%T)", decAttr, decAttr)
 	}
 	opType := p.typeMap[expr.GetCallExpr().GetTarget().GetId()]
 	indConst, isIndConst := ind.(InterpretableConst)
@@ -513,8 +502,11 @@ func (p *planner) planCallIndex(expr *exprpb.Expr,
 		_, err = opAttr.AddQualifier(qual)
 		return opAttr, err
 	}
-	_, err = opAttr.AddQualifier(
-		p.attrFactory.RelativeAttribute(expr.GetId(), ind))
+	indQual, err := p.relativeAttr(expr.GetId(), ind)
+	if err != nil {
+		return nil, err
+	}
+	_, err = opAttr.AddQualifier(indQual)
 	return opAttr, err
 }
 
@@ -738,6 +730,25 @@ func (p *planner) resolveFunction(expr *exprpb.Expr) (*exprpb.Expr, string, stri
 	// In the default case, the function is exactly as it was advertised: a receiver call on with
 	// an expression-based target with the given simple function name.
 	return target, fnName, ""
+}
+
+func (p *planner) relativeAttr(id int64, eval Interpretable) (InterpretableAttribute, error) {
+	eAttr, ok := eval.(InterpretableAttribute)
+	if !ok {
+		eAttr = &evalAttr{
+			adapter: p.adapter,
+			attr:    p.attrFactory.RelativeAttribute(id, eval),
+		}
+	}
+	decAttr, err := p.decorate(eAttr, nil)
+	if err != nil {
+		return nil, err
+	}
+	eAttr, ok = decAttr.(InterpretableAttribute)
+	if !ok {
+		return nil, fmt.Errorf("invalid attribute decoration: %v(%T)", decAttr, decAttr)
+	}
+	return eAttr, nil
 }
 
 // toQualifiedName converts an expression AST into a qualified name if possible, with a boolean

--- a/interpreter/prune.go
+++ b/interpreter/prune.go
@@ -105,8 +105,7 @@ func (p *astPruner) maybeCreateLiteral(id int64, val ref.Val) (*exprpb.Expr, boo
 	}
 
 	// Attempt to build a list literal.
-	list, isList := val.(traits.Lister)
-	if isList {
+	if list, isList := val.(traits.Lister); isList {
 		sz := list.Size().(types.Int)
 		elemExprs := make([]*exprpb.Expr, sz)
 		for i := types.Int(0); i < sz; i++ {
@@ -131,8 +130,7 @@ func (p *astPruner) maybeCreateLiteral(id int64, val ref.Val) (*exprpb.Expr, boo
 	}
 
 	// Create a map literal if possible.
-	mp, isMap := val.(traits.Mapper)
-	if isMap {
+	if mp, isMap := val.(traits.Mapper); isMap {
 		it := mp.Iterator()
 		entries := make([]*exprpb.Expr_CreateStruct_Entry, mp.Size().(types.Int))
 		i := 0
@@ -383,10 +381,12 @@ func (p *astPruner) existsWithKnownValue(id int64) bool {
 }
 
 func (p *astPruner) nextID() int64 {
-	_, found := p.state.Value(p.nextExprID)
-	for found {
+	for true {
+		_, found := p.state.Value(p.nextExprID)
+		if found {
+			break
+		}
 		p.nextExprID++
-		_, found = p.state.Value(p.nextExprID)
 	}
 	next := p.nextExprID
 	p.nextExprID++

--- a/interpreter/prune.go
+++ b/interpreter/prune.go
@@ -18,15 +18,18 @@ import (
 	"github.com/google/cel-go/common/operators"
 	"github.com/google/cel-go/common/types"
 	"github.com/google/cel-go/common/types/ref"
+	"github.com/google/cel-go/common/types/traits"
 
 	structpb "github.com/golang/protobuf/ptypes/struct"
 
+	expr "google.golang.org/genproto/googleapis/api/expr/v1alpha1"
 	exprpb "google.golang.org/genproto/googleapis/api/expr/v1alpha1"
 )
 
 type astPruner struct {
-	expr  *exprpb.Expr
-	state EvalState
+	expr       *exprpb.Expr
+	state      EvalState
+	nextExprID int64
 }
 
 // TODO Consider having a separate walk of the AST that finds common
@@ -60,19 +63,116 @@ type astPruner struct {
 // the overloads accordingly.
 func PruneAst(expr *exprpb.Expr, state EvalState) *exprpb.Expr {
 	pruner := &astPruner{
-		expr:  expr,
-		state: state}
+		expr:       expr,
+		state:      state,
+		nextExprID: 1}
 	newExpr, _ := pruner.prune(expr)
 	return newExpr
 }
 
-func (p *astPruner) createLiteral(node *exprpb.Expr, val *exprpb.Constant) *exprpb.Expr {
+func (p *astPruner) createLiteral(id int64, val *exprpb.Constant) *exprpb.Expr {
 	return &exprpb.Expr{
-		Id: node.GetId(),
+		Id: id,
 		ExprKind: &exprpb.Expr_ConstExpr{
 			ConstExpr: val,
 		},
 	}
+}
+
+func (p *astPruner) maybeCreateLiteral(id int64, val ref.Val) (*exprpb.Expr, bool) {
+	switch val.Type() {
+	case types.BoolType:
+		return p.createLiteral(id,
+			&exprpb.Constant{ConstantKind: &exprpb.Constant_BoolValue{BoolValue: val.Value().(bool)}}), true
+	case types.IntType:
+		return p.createLiteral(id,
+			&exprpb.Constant{ConstantKind: &exprpb.Constant_Int64Value{Int64Value: val.Value().(int64)}}), true
+	case types.UintType:
+		return p.createLiteral(id,
+			&exprpb.Constant{ConstantKind: &exprpb.Constant_Uint64Value{Uint64Value: val.Value().(uint64)}}), true
+	case types.StringType:
+		return p.createLiteral(id,
+			&exprpb.Constant{ConstantKind: &exprpb.Constant_StringValue{StringValue: val.Value().(string)}}), true
+	case types.DoubleType:
+		return p.createLiteral(id,
+			&exprpb.Constant{ConstantKind: &exprpb.Constant_DoubleValue{DoubleValue: val.Value().(float64)}}), true
+	case types.BytesType:
+		return p.createLiteral(id,
+			&exprpb.Constant{ConstantKind: &exprpb.Constant_BytesValue{BytesValue: val.Value().([]byte)}}), true
+	case types.NullType:
+		return p.createLiteral(id,
+			&exprpb.Constant{ConstantKind: &exprpb.Constant_NullValue{NullValue: val.Value().(structpb.NullValue)}}), true
+	}
+
+	// Attempt to build a list literal.
+	list, isList := val.(traits.Lister)
+	if isList {
+		sz := list.Size().(types.Int)
+		elemExprs := make([]*exprpb.Expr, sz)
+		for i := types.Int(0); i < sz; i++ {
+			elem := list.Get(i)
+			if types.IsUnknownOrError(elem) {
+				return nil, false
+			}
+			elemExpr, ok := p.maybeCreateLiteral(p.nextID(), elem)
+			if !ok {
+				return nil, false
+			}
+			elemExprs[i] = elemExpr
+		}
+		return &exprpb.Expr{
+			Id: id,
+			ExprKind: &exprpb.Expr_ListExpr{
+				ListExpr: &expr.Expr_CreateList{
+					Elements: elemExprs,
+				},
+			},
+		}, true
+	}
+
+	// Create a map literal if possible.
+	mp, isMap := val.(traits.Mapper)
+	if isMap {
+		it := mp.Iterator()
+		entries := make([]*exprpb.Expr_CreateStruct_Entry, mp.Size().(types.Int))
+		i := 0
+		for it.HasNext() != types.False {
+			key := it.Next()
+			val := mp.Get(key)
+			if types.IsUnknownOrError(key) || types.IsUnknownOrError(val) {
+				return nil, false
+			}
+			keyExpr, ok := p.maybeCreateLiteral(p.nextID(), key)
+			if !ok {
+				return nil, false
+			}
+			valExpr, ok := p.maybeCreateLiteral(p.nextID(), val)
+			if !ok {
+				return nil, false
+			}
+			entry := &exprpb.Expr_CreateStruct_Entry{
+				Id: p.nextID(),
+				KeyKind: &exprpb.Expr_CreateStruct_Entry_MapKey{
+					MapKey: keyExpr,
+				},
+				Value: valExpr,
+			}
+			entries[i] = entry
+			i++
+		}
+		return &exprpb.Expr{
+			Id: id,
+			ExprKind: &exprpb.Expr_StructExpr{
+				StructExpr: &exprpb.Expr_CreateStruct{
+					Entries: entries,
+				},
+			},
+		}, true
+	}
+
+	// TODO(issues/377) To construct message literals, the type provider will need to support
+	// the enumeration the fields for a given message.
+	return nil, false
 }
 
 func (p *astPruner) maybePruneAndOr(node *exprpb.Expr) (*exprpb.Expr, bool) {
@@ -125,32 +225,10 @@ func (p *astPruner) prune(node *exprpb.Expr) (*exprpb.Expr, bool) {
 	if node == nil {
 		return node, false
 	}
-	if val, valueExists := p.value(node.GetId()); valueExists && !types.IsUnknownOrError(val) {
-		// TODO if we have a list or struct, create a list/struct
-		// expression. This is useful especially if these expressions
-		// are result of a function call.
-		switch val.Type() {
-		case types.BoolType:
-			return p.createLiteral(node,
-				&exprpb.Constant{ConstantKind: &exprpb.Constant_BoolValue{BoolValue: val.Value().(bool)}}), true
-		case types.IntType:
-			return p.createLiteral(node,
-				&exprpb.Constant{ConstantKind: &exprpb.Constant_Int64Value{Int64Value: val.Value().(int64)}}), true
-		case types.UintType:
-			return p.createLiteral(node,
-				&exprpb.Constant{ConstantKind: &exprpb.Constant_Uint64Value{Uint64Value: val.Value().(uint64)}}), true
-		case types.StringType:
-			return p.createLiteral(node,
-				&exprpb.Constant{ConstantKind: &exprpb.Constant_StringValue{StringValue: val.Value().(string)}}), true
-		case types.DoubleType:
-			return p.createLiteral(node,
-				&exprpb.Constant{ConstantKind: &exprpb.Constant_DoubleValue{DoubleValue: val.Value().(float64)}}), true
-		case types.BytesType:
-			return p.createLiteral(node,
-				&exprpb.Constant{ConstantKind: &exprpb.Constant_BytesValue{BytesValue: val.Value().([]byte)}}), true
-		case types.NullType:
-			return p.createLiteral(node,
-				&exprpb.Constant{ConstantKind: &exprpb.Constant_NullValue{NullValue: val.Value().(structpb.NullValue)}}), true
+	val, valueExists := p.value(node.GetId())
+	if valueExists && !types.IsUnknownOrError(val) {
+		if newNode, ok := p.maybeCreateLiteral(node.GetId(), val); ok {
+			return newNode, true
 		}
 	}
 
@@ -187,7 +265,7 @@ func (p *astPruner) prune(node *exprpb.Expr) (*exprpb.Expr, bool) {
 			Args:     newArgs,
 		}
 		for i, arg := range args {
-			newArgs[i] = args[i]
+			newArgs[i] = arg
 			if newArg, prunedArg := p.prune(arg); prunedArg {
 				prunedCall = true
 				newArgs[i] = newArg
@@ -210,6 +288,7 @@ func (p *astPruner) prune(node *exprpb.Expr) (*exprpb.Expr, bool) {
 		newElems := make([]*exprpb.Expr, len(elems))
 		var prunedList bool
 		for i, elem := range elems {
+			newElems[i] = elem
 			if newElem, prunedElem := p.prune(elem); prunedElem {
 				newElems[i] = newElem
 				prunedList = true
@@ -231,10 +310,11 @@ func (p *astPruner) prune(node *exprpb.Expr) (*exprpb.Expr, bool) {
 		messageType := node.GetStructExpr().GetMessageName()
 		newEntries := make([]*exprpb.Expr_CreateStruct_Entry, len(entries))
 		for i, entry := range entries {
+			newEntries[i] = entry
 			newKey, prunedKey := p.prune(entry.GetMapKey())
 			newValue, prunedValue := p.prune(entry.GetValue())
 			if !prunedKey && !prunedValue {
-				newEntries[i] = entry
+				continue
 			}
 			prunedStruct = true
 			newEntry := &exprpb.Expr_CreateStruct_Entry{
@@ -264,6 +344,9 @@ func (p *astPruner) prune(node *exprpb.Expr) (*exprpb.Expr, bool) {
 		}
 	case *exprpb.Expr_ComprehensionExpr:
 		compre := node.GetComprehensionExpr()
+		// Only the range of the comprehension is pruned since the state tracking only records
+		// the last iteration of the comprehension and not each step in the evaluation which
+		// means that the any residuals computed in between might be inaccurate.
 		if newRange, pruned := p.prune(compre.GetIterRange()); pruned {
 			return &exprpb.Expr{
 				Id: node.GetId(),
@@ -297,4 +380,15 @@ func (p *astPruner) existsWithUnknownValue(id int64) bool {
 func (p *astPruner) existsWithKnownValue(id int64) bool {
 	val, valueExists := p.value(id)
 	return valueExists && !types.IsUnknown(val)
+}
+
+func (p *astPruner) nextID() int64 {
+	_, found := p.state.Value(p.nextExprID)
+	for found {
+		p.nextExprID++
+		_, found = p.state.Value(p.nextExprID)
+	}
+	next := p.nextExprID
+	p.nextExprID++
+	return next
 }

--- a/interpreter/prune.go
+++ b/interpreter/prune.go
@@ -67,9 +67,12 @@ func PruneAst(expr *exprpb.Expr, state EvalState) *exprpb.Expr {
 }
 
 func (p *astPruner) createLiteral(node *exprpb.Expr, val *exprpb.Constant) *exprpb.Expr {
-	newExpr := *node
-	newExpr.ExprKind = &exprpb.Expr_ConstExpr{ConstExpr: val}
-	return &newExpr
+	return &exprpb.Expr{
+		Id: node.GetId(),
+		ExprKind: &exprpb.Expr_ConstExpr{
+			ConstExpr: val,
+		},
+	}
 }
 
 func (p *astPruner) maybePruneAndOr(node *exprpb.Expr) (*exprpb.Expr, bool) {
@@ -78,13 +81,11 @@ func (p *astPruner) maybePruneAndOr(node *exprpb.Expr) (*exprpb.Expr, bool) {
 	}
 
 	call := node.GetCallExpr()
-
 	// We know result is unknown, so we have at least one unknown arg
 	// and if one side is a known value, we know we can ignore it.
 	if p.existsWithKnownValue(call.Args[0].GetId()) {
 		return call.Args[1], true
 	}
-
 	if p.existsWithKnownValue(call.Args[1].GetId()) {
 		return call.Args[0], true
 	}
@@ -125,11 +126,9 @@ func (p *astPruner) prune(node *exprpb.Expr) (*exprpb.Expr, bool) {
 		return node, false
 	}
 	if val, valueExists := p.value(node.GetId()); valueExists && !types.IsUnknownOrError(val) {
-
 		// TODO if we have a list or struct, create a list/struct
 		// expression. This is useful especially if these expressions
 		// are result of a function call.
-
 		switch val.Type() {
 		case types.BoolType:
 			return p.createLiteral(node,
@@ -162,79 +161,124 @@ func (p *astPruner) prune(node *exprpb.Expr) (*exprpb.Expr, bool) {
 	switch node.ExprKind.(type) {
 	case *exprpb.Expr_SelectExpr:
 		if operand, pruned := p.prune(node.GetSelectExpr().Operand); pruned {
-			newExpr := *node
-			newSelect := *newExpr.GetSelectExpr()
-			newSelect.Operand = operand
-			newExpr.GetExprKind().(*exprpb.Expr_SelectExpr).SelectExpr = &newSelect
-			return &newExpr, true
+			return &exprpb.Expr{
+				Id: node.GetId(),
+				ExprKind: &exprpb.Expr_SelectExpr{
+					SelectExpr: &exprpb.Expr_Select{
+						Operand:  operand,
+						Field:    node.GetSelectExpr().GetField(),
+						TestOnly: node.GetSelectExpr().GetTestOnly(),
+					},
+				},
+			}, true
 		}
 	case *exprpb.Expr_CallExpr:
 		if newExpr, pruned := p.maybePruneFunction(node); pruned {
 			newExpr, _ = p.prune(newExpr)
 			return newExpr, true
 		}
-		newCall := *node.GetCallExpr()
 		var prunedCall bool
-		var prunedArg bool
-		for i, arg := range node.GetCallExpr().Args {
-			if newCall.Args[i], prunedArg = p.prune(arg); prunedArg {
+		call := node.GetCallExpr()
+		args := call.GetArgs()
+		newArgs := make([]*exprpb.Expr, len(args))
+		newCall := &exprpb.Expr_Call{
+			Function: call.GetFunction(),
+			Target:   call.GetTarget(),
+			Args:     newArgs,
+		}
+		for i, arg := range args {
+			newArgs[i] = args[i]
+			if newArg, prunedArg := p.prune(arg); prunedArg {
 				prunedCall = true
+				newArgs[i] = newArg
 			}
 		}
-		if newTarget, prunedTarget := p.prune(node.GetCallExpr().Target); prunedTarget {
+		if newTarget, prunedTarget := p.prune(call.GetTarget()); prunedTarget {
 			prunedCall = true
 			newCall.Target = newTarget
 		}
 		if prunedCall {
-			newExpr := *node
-			newExpr.GetExprKind().(*exprpb.Expr_CallExpr).CallExpr = &newCall
-			return &newExpr, true
+			return &exprpb.Expr{
+				Id: node.GetId(),
+				ExprKind: &exprpb.Expr_CallExpr{
+					CallExpr: newCall,
+				},
+			}, true
 		}
 	case *exprpb.Expr_ListExpr:
-		newList := *node.GetListExpr()
+		elems := node.GetListExpr().GetElements()
+		newElems := make([]*exprpb.Expr, len(elems))
 		var prunedList bool
-		var prunedElem bool
-		for i, elem := range node.GetListExpr().Elements {
-			if newList.Elements[i], prunedElem = p.prune(elem); prunedElem {
+		for i, elem := range elems {
+			if newElem, prunedElem := p.prune(elem); prunedElem {
+				newElems[i] = newElem
 				prunedList = true
 			}
 		}
 		if prunedList {
-			newExpr := *node
-			newExpr.GetExprKind().(*exprpb.Expr_ListExpr).ListExpr = &newList
-			return &newExpr, true
+			return &exprpb.Expr{
+				Id: node.GetId(),
+				ExprKind: &exprpb.Expr_ListExpr{
+					ListExpr: &exprpb.Expr_CreateList{
+						Elements: newElems,
+					},
+				},
+			}, true
 		}
 	case *exprpb.Expr_StructExpr:
-		newStruct := *node.GetStructExpr()
 		var prunedStruct bool
-		var prunedEntry bool
-		for i, entry := range node.GetStructExpr().Entries {
-			newEntry := *entry
-			if newKey, pruned := p.prune(entry.GetMapKey()); pruned {
-				prunedEntry = true
-				newEntry.GetKeyKind().(*exprpb.Expr_CreateStruct_Entry_MapKey).MapKey = newKey
+		entries := node.GetStructExpr().GetEntries()
+		messageType := node.GetStructExpr().GetMessageName()
+		newEntries := make([]*exprpb.Expr_CreateStruct_Entry, len(entries))
+		for i, entry := range entries {
+			newKey, prunedKey := p.prune(entry.GetMapKey())
+			newValue, prunedValue := p.prune(entry.GetValue())
+			if !prunedKey && !prunedValue {
+				newEntries[i] = entry
 			}
-			if newValue, pruned := p.prune(entry.Value); pruned {
-				prunedEntry = true
-				newEntry.Value = newValue
+			prunedStruct = true
+			newEntry := &exprpb.Expr_CreateStruct_Entry{
+				Value: newValue,
 			}
-			if prunedEntry {
-				prunedStruct = true
-				newStruct.Entries[i] = &newEntry
+			if messageType != "" {
+				newEntry.KeyKind = &exprpb.Expr_CreateStruct_Entry_FieldKey{
+					FieldKey: entry.GetFieldKey(),
+				}
+			} else {
+				newEntry.KeyKind = &exprpb.Expr_CreateStruct_Entry_MapKey{
+					MapKey: newKey,
+				}
 			}
+			newEntries[i] = newEntry
 		}
 		if prunedStruct {
-			newExpr := *node
-			newExpr.GetExprKind().(*exprpb.Expr_StructExpr).StructExpr = &newStruct
-			return &newExpr, true
+			return &exprpb.Expr{
+				Id: node.GetId(),
+				ExprKind: &exprpb.Expr_StructExpr{
+					StructExpr: &exprpb.Expr_CreateStruct{
+						MessageName: messageType,
+						Entries:     newEntries,
+					},
+				},
+			}, true
 		}
 	case *exprpb.Expr_ComprehensionExpr:
-		if newIterRange, pruned := p.prune(node.GetComprehensionExpr().IterRange); pruned {
-			newExpr := *node
-			newCompre := *newExpr.GetComprehensionExpr()
-			newCompre.IterRange = newIterRange
-			newExpr.GetExprKind().(*exprpb.Expr_ComprehensionExpr).ComprehensionExpr = &newCompre
-			return &newExpr, true
+		compre := node.GetComprehensionExpr()
+		if newRange, pruned := p.prune(compre.GetIterRange()); pruned {
+			return &exprpb.Expr{
+				Id: node.GetId(),
+				ExprKind: &exprpb.Expr_ComprehensionExpr{
+					ComprehensionExpr: &exprpb.Expr_Comprehension{
+						IterVar:       compre.GetIterVar(),
+						IterRange:     newRange,
+						AccuVar:       compre.GetAccuVar(),
+						AccuInit:      compre.GetAccuInit(),
+						LoopCondition: compre.GetLoopCondition(),
+						LoopStep:      compre.GetLoopStep(),
+						Result:        compre.GetResult(),
+					},
+				},
+			}, true
 		}
 	}
 	return node, false

--- a/interpreter/prune_test.go
+++ b/interpreter/prune_test.go
@@ -17,126 +17,129 @@ package interpreter
 import (
 	"testing"
 
-	"github.com/google/cel-go/common/debug"
-	"github.com/google/cel-go/common/operators"
+	"github.com/google/cel-go/common"
 	"github.com/google/cel-go/common/containers"
 	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/parser"
 	"github.com/google/cel-go/test"
-
-	exprpb "google.golang.org/genproto/googleapis/api/expr/v1alpha1"
 )
 
 type testInfo struct {
-	A Activation
-	E *exprpb.Expr
-	P string
+	in   Activation
+	expr string
+	out  string
 }
 
 var testCases = []testInfo{
 	{
-		A: unknownActivation(),
-		E: test.ExprCall(2, operators.LogicalAnd,
-			test.ExprLiteral(1, true),
-			test.ExprLiteral(3, false)),
-		P: `false`,
+		expr: `true && false`,
+		out:  `false`,
 	},
 	{
-		A: unknownActivation("x"),
-		E: test.ExprCall(4, operators.LogicalAnd,
-			test.ExprCall(2, operators.LogicalOr,
-				test.ExprLiteral(1, true),
-				test.ExprLiteral(3, false)),
-			test.ExprIdent(5, "x")),
-		P: `x`,
+		in:   unknownActivation("x"),
+		expr: `(true || false) && x`,
+		out:  `x`,
 	},
 	{
-		A: unknownActivation("x"),
-		E: test.ExprCall(4, operators.LogicalAnd,
-			test.ExprCall(2, operators.LogicalOr,
-				test.ExprLiteral(1, false),
-				test.ExprLiteral(3, false)),
-			test.ExprIdent(5, "x")),
-		P: `false`,
+		in:   unknownActivation("x"),
+		expr: `(false || false) && x`,
+		out:  `false`,
 	},
 	{
-		A: unknownActivation("a"),
-		E: test.ExprCall(2, operators.LogicalAnd,
-			test.ExprIdent(1, "a"),
-			test.ExprComprehension(3,
-				"x",
-				test.ExprList(7,
-					test.ExprLiteral(4, int64(1)),
-					test.ExprLiteral(5, uint64(1)),
-					test.ExprLiteral(6, float64(1.0))),
-				"__result__",
-				test.ExprLiteral(8, false),
-				test.ExprCall(11,
-					operators.NotStrictlyFalse,
-					test.ExprCall(9,
-						operators.LogicalNot,
-						test.ExprIdent(10, "__result__"))),
-				test.ExprCall(12,
-					operators.LogicalOr,
-					test.ExprIdent(13, "__result__"),
-					test.ExprCall(14,
-						operators.Equals,
-						test.ExprCall(15,
-							"type",
-							test.ExprIdent(16, "x")),
-						test.ExprIdent(17, "uint"))),
-				test.ExprIdent(18, "__result__"))),
-		P: `a`,
+		in:   unknownActivation("a"),
+		expr: `a && [1, 1u, 1.0].exists(x, type(x) == uint)`,
+		out:  `a`,
 	},
 	{
-		A: unknownActivation(),
-		E: test.ExprMap(8,
-			test.ExprEntry(2,
-				test.ExprLiteral(1, "hello"),
-				test.ExprMemberCall(3,
-					"size",
-					test.ExprLiteral(4, "world"))),
-			test.ExprEntry(6,
-				test.ExprLiteral(5, "bytes"),
-				test.ExprLiteral(7, []byte("bytes-string")))),
-		P: `{"hello":5, "bytes":b"bytes-string"}`,
+		expr: `{'hello': 'world'.size()}`,
+		out:  `{"hello": 5}`,
 	},
 	{
-		E: test.ExprCall(1, operators.Less,
-			test.ExprLiteral(2, int64(2)),
-			test.ExprLiteral(3, int64(3))),
-		P: `true`,
+		expr: `[b'bytes-string']`,
+		out:  `[b"\142\171\164\145\163\055\163\164\162\151\156\147"]`,
 	},
 	{
-		A: unknownActivation("b", "c"),
-		E: test.ExprCall(8, operators.Conditional,
-			test.ExprLiteral(1, true),
-			test.ExprCall(3,
-				operators.Less,
-				test.ExprIdent(2, "b"),
-				test.ExprLiteral(4, 1.2)),
-			test.ExprCall(6,
-				operators.Equals,
-				test.ExprIdent(5, "c"),
-				test.ExprList(7, test.ExprLiteral(7, "hello")))),
-		P: `_<_(b,1.2)`,
+		expr: `[b'bytes'] + [b'-' + b'string']`,
+		out:  `[b"\142\171\164\145\163", b"\055\163\164\162\151\156\147"]`,
 	},
+	{
+		expr: `1u + 3u`,
+		out:  `4u`,
+	},
+	{
+		expr: `2 < 3`,
+		out:  `true`,
+	},
+	{
+		in:   unknownActivation(),
+		expr: `test == null`,
+		out:  `test == null`,
+	},
+	{
+		in:   unknownActivation(),
+		expr: `test == null && false`,
+		out:  `false`,
+	},
+	{
+		in:   unknownActivation("b", "c"),
+		expr: `true ? b < 1.2 : c == ['hello']`,
+		out:  `b < 1.2`,
+	},
+	{
+		in:   unknownActivation(),
+		expr: `[1+3, 2+2, 3+1, four]`,
+		out:  `[4, 4, 4, four]`,
+	},
+	{
+		in:   unknownActivation(),
+		expr: `test == {'a': 1, 'field': 2}.field`,
+		out:  `test == 2`,
+	},
+	{
+		in:   unknownActivation(),
+		expr: `test in {'a': 1, 'field': [2, 3]}.field`,
+		out:  `test in [2, 3]`,
+	},
+	{
+		in:   unknownActivation(),
+		expr: `test == {'field': [1 + 2, 2 + 3]}`,
+		out:  `test == {"field": [3, 5]}`,
+	},
+	{
+		in:   unknownActivation(),
+		expr: `test in {'a': 1, 'field': [test, 3]}.field`,
+		out:  `test in {"a": 1, "field": [test, 3]}.field`,
+	},
+	// TODO(issues/) the output test relies on tracking macro expansions back to their original
+	// call patterns.
+	/* {
+		in:   unknownActivation(),
+		expr: `[1+3, 2+2, 3+1, four].exists(x, x == four)`,
+		out:  `[4, 4, 4, four].exists(x, x == four)`,
+	}, */
 }
 
 func TestPrune(t *testing.T) {
 	for i, tst := range testCases {
-		pExpr := &exprpb.ParsedExpr{Expr: tst.E}
+		ast, iss := parser.Parse(common.NewStringSource(tst.expr, "<input>"))
+		if len(iss.GetErrors()) > 0 {
+			t.Fatalf(iss.ToDisplayString())
+		}
 		state := NewEvalState()
 		reg := types.NewRegistry()
 		attrs := NewPartialAttributeFactory(containers.DefaultContainer, reg, reg)
 		interp := NewStandardInterpreter(containers.DefaultContainer, reg, reg, attrs)
 		interpretable, _ := interp.NewUncheckedInterpretable(
-			pExpr.Expr,
+			ast.Expr,
 			ExhaustiveEval(state))
-		interpretable.Eval(tst.A)
-		newExpr := PruneAst(pExpr.Expr, state)
-		actual := debug.ToDebugString(newExpr)
-		if !test.Compare(actual, tst.P) {
-			t.Errorf("prune[%d], diff: %s", i, test.DiffMessage("structure", actual, tst.P))
+		interpretable.Eval(tst.in)
+		newExpr := PruneAst(ast.Expr, state)
+		actual, err := parser.Unparse(newExpr, nil)
+		if err != nil {
+			t.Error(err)
+		}
+		if !test.Compare(actual, tst.out) {
+			t.Errorf("prune[%d], diff: %s", i, test.DiffMessage("structure", actual, tst.out))
 		}
 	}
 }

